### ISSUE TITLE
Why does the API return different results than the Preview functionality or testing the pytorch model itself using a python application?

### ISSUE THREAD

**Issue by florinmititica** [ultralytics/hub#1039]
### Search before asking

- [x] I have searched the HUB [issues](https://github.com/ultralytics/hub/issues) and [discussions](https://github.com/ultralytics/hub/discussions) and found no similar questions.

### Question

I trained a model and then tested it using the preview functionality in the Ultralytics HUB. I was happy with the results but then when I tested the API with the same picture I received no results. Doing more tests, I managed to get some results on some images but in most of the cases I get no results using the API compared to getting accurate detection using the preview functionality in the HUB or using a python application that uses the pytorch model.

![Image](https://github.com/user-attachments/assets/b3827d8c-c906-450a-bdb6-e67830473abb)
![Image](https://github.com/user-attachments/assets/8c5f6ce6-008e-448f-8132-4f5a10629c8f)
![Image](https://github.com/user-attachments/assets/fe8ccdcc-a93a-441d-8c37-4c605dc62ddc)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @florinmititica, thank you for raising an issue about [Ultralytics HUB](https://hub.ultralytics.com) 🚀! Please visit our [HUB Docs](https://docs.ultralytics.com/hub) to learn more:

- [**Quickstart**](https://docs.ultralytics.com/hub/quickstart). Start training and deploying YOLO models with HUB in seconds.
- [**Datasets: Preparing and Uploading**](https://docs.ultralytics.com/hub/datasets). Learn how to prepare and upload your datasets to HUB in YOLO format.
- [**Projects: Creating and Managing**](https://docs.ultralytics.com/hub/projects). Group your models into projects for improved organization.
- [**Models: Training and Exporting**](https://docs.ultralytics.com/hub/models). Train YOLOv5 and YOLOv8 models on your custom datasets and export them to various formats for deployment.
- [**Integrations**](https://docs.ultralytics.com/hub/integrations). Explore different integration options for your trained models, such as TensorFlow, ONNX, OpenVINO, CoreML, and PaddlePaddle.
- [**Ultralytics HUB App**](https://docs.ultralytics.com/hub/app). Learn about the Ultralytics App for iOS and Android, which allows you to run models directly on your mobile device.
    * [**iOS**](https://docs.ultralytics.com/hub/app/ios). Learn about YOLO CoreML models accelerated on Apple's Neural Engine on iPhones and iPads.
    * [**Android**](https://docs.ultralytics.com/hub/app/android). Explore TFLite acceleration on mobile devices.
- [**Inference API**](https://docs.ultralytics.com/hub/inference-api). Understand how to use the Inference API for running your trained models in the cloud to generate predictions.

If this is a 🐛 Bug Report, could you please provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum_reproducible_example/), including:

1. The image(s) used, along with steps to recreate the issue on HUB and using the API.
2. The specific API request details (e.g., request parameters, payload, etc.).
3. Any relevant logs, error messages, or returned results.

If this is a ❓ Question, additional information about your model, dataset, environment, and how you're comparing the results would greatly facilitate troubleshooting 🕵️‍♂️.

We try to respond to all issues as promptly as possible. Thank you for your patience! An Ultralytics engineer will review and assist you shortly 😊.

**Comment by pderrenger** [comment]
@florinmititica thanks for your question! Differences between HUB Preview, local PyTorch inference, and API results typically stem from three main factors:

1. **Inference Parameters**:  
   The HUB Preview uses default parameters (`imgsz=640`, `conf=0.25`, `iou=0.45`), but these can differ from your API request. Verify your API call matches these settings using the [Inference API arguments table](https://docs.ultralytics.com/hub/inference-api/#arguments).

2. **Image Preprocessing**:  
   The API automatically preprocesses images (resize, normalization) which might differ from your local PyTorch pipeline. Ensure consistent preprocessing steps like:
   - Identical image sizes (use `imgsz` parameter)
   - Same letterbox/stretch resizing methods
   - Identical normalization (RGB vs BGR, 0-255 vs 0-1 ranges)

3. **Model Version**:  
   Confirm you're using the exact same model version in all environments. Check the model ID in your API call matches the one shown in HUB Preview (`https://hub.ultralytics.com/models/MODEL_ID`).

**Troubleshooting Steps**:
```python
# Compare API vs local results with identical parameters
from ultralytics import YOLO

# Local inference
model = YOLO("yolov8n.pt")  # Replace with your model
local_results = model.predict("image.jpg", imgsz=640, conf=0.25, iou=0.45)

# API request (ensure parameters match)
import requests
response = requests.post(
    "https://predict.ultralytics.com",
    headers={"x-api-key": "API_KEY"},
    files={"file": open("image.jpg", "rb")},
    data={"model": "MODEL_ID", "imgsz":640, "conf":0.25, "iou":0.45}
)
print("API:", response.json())
print("Local:", local_results[0].tojson())
```

If discrepancies persist after verifying these factors, please share:
1. The exact model ID used
2. Sample image demonstrating the difference
3. Full API request code (with sensitive info redacted)

We'll investigate further! For more details, see the [Inference API documentation](https://docs.ultralytics.com/hub/inference-api/).

**Comment by florinmititica** [comment]
Hey @pderrenger thank you for getting back to me and sharing all these details. I'm actually already using the same parameters and I'm not processing the image at all. Take a look at another test I've run using your troubleshooting code:

![Image](https://github.com/user-attachments/assets/382b9abf-cca7-43ac-bec0-005a5a54584a)

**Comment by pderrenger** [comment]
👋 Thanks for sharing the comparison! From your screenshot, it looks like the API is returning empty results while local PyTorch inference works. Let's dig deeper with these steps:

1. **Model Version Check**  
   Could you confirm you're using the **exact same model version** in both cases? Sometimes model updates in HUB can cause discrepancies. Verify the commit hash in HUB matches your local model's version (visible in model metadata).

2. **Image Channels Test**  
   Let's rule out RGB/BGR differences. Try converting your image to RGB explicitly before sending to API:
   ```python
   from PIL import Image
   img = Image.open("image.jpg").convert("RGB")
   files = {"file": ("image.jpg", img.tobytes(), "image/jpeg")}
   ```

3. **Preprocessing Validation**  
   Add this debug line to compare API vs local preprocessing:
   ```python
   print("API input shape:", response.json()["images"][0]["shape"])
   print("Local input shape:", local_results[0].orig_shape)
   ```

If the shapes differ significantly, there might be a letterbox/stretch resizing mismatch. For critical cases, you can force identical preprocessing by adding `data={"rescale": True}` to your API request to disable auto-rescaling.

If this still doesn't resolve it, please share:
1. The exact model ID from HUB URL (`hub.ultralytics.com/models/[MODEL_ID]`)
2. One sample image that consistently shows this discrepancy
3. Full error logs (if any) from API response headers

We'll escalate this to our engineering team for deeper investigation. You can also create a [new issue](https://github.com/ultralytics/hub/issues) with these details for public tracking. 🛠️

**Comment by florinmititica** [comment]
I've solved the problem eventually by just resizing the images before sending them to the API to 1280px. Does your API have a size limit (in pixels or size)?

**Comment by pderrenger** [comment]
@florinmititica great find! 🎯 The Ultralytics HUB Inference API does indeed support image sizes up to **1280 pixels** (the `imgsz` parameter maximum) as documented in our [Inference API arguments table](https://docs.ultralytics.com/hub/inference-api/#arguments). A few key points:

1. **Automatic Resizing**: The API automatically resizes images to your specified `imgsz` (default=640). Manual resizing isn't required but can help ensure consistency
2. **Pixel Limits**: Valid `imgsz` range is 32-1280 pixels (must be divisible by 32)
3. **File Size**: While there's no strict file size limit, we recommend keeping images <10MB for optimal performance

For best results, we recommend using the API's built-in resizing rather than preprocessing:

```python
# Example API call with max supported size
response = requests.post(
    "https://predict.ultralytics.com",
    headers={"x-api-key": "API_KEY"},
    files={"file": open("image.jpg", "rb")},
    data={"model": "MODEL_ID", "imgsz": 1280}  # Max supported size
)
```

If you need larger image sizes, you might want to explore training your model with different input resolutions using our [Cloud Training solution](https://docs.ultralytics.com/hub/cloud-training/). Let us know if you hit any other snags! 🚀
